handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java

```diff
@@ -0,0 +1,360 @@
+                       timeoutGotLatch.await(5 * TIMEOUT_MS, TimeUnit.MILLISECONDS));
+            assertTrue("ReadAfterWriteTimeoutHandler should close channel by default",
+                       clientClosedChannelLatch.await(3 * TIMEOUT_MS, TimeUnit.MILLISECONDS));
+            assertFalse("Unexpected exception in server handler", unexpectedExceptionInServerHandler.get());
+            assertFalse("Unexpected exception in client handler", unexpectedExceptionInClientHandler.get());
+        } finally {
+            client.close();
+            server.close();
+        }
+    }
+
+    @Test
+    public void testOnlyOneTimeoutProducedForMultipleConsecutiveWrites() throws Exception {
+        final int TIMEOUT_MS = 300;
+        final int MESSAGES_COUNT = 2;
+
+        final AtomicInteger serverReads = new AtomicInteger();
+        final CountDownLatch timeoutGotLatch = new CountDownLatch(MESSAGES_COUNT);
+        final AtomicBoolean unexpectedExceptionInServerHandler = new AtomicBoolean();
+        final AtomicBoolean unexpectedExceptionInClientHandler = new AtomicBoolean();
+
+        ReadAfterWriteTimeoutHandler readAfterWriteTimeoutHandler =
+                new ReadAfterWriteTimeoutHandler(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+
+        ChannelInboundHandler pongAfterTwoMessagesHandler = new ChannelInboundHandlerAdapter() {
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                serverReads.incrementAndGet();
+                //response once only after MESSAGES_COUNT messages and sleeping
+                if (serverReads.get() == MESSAGES_COUNT) {
+                    Thread.sleep(2 * TIMEOUT_MS);
+                    ctx.writeAndFlush(newOneMessage());
+                }
+            }
+
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                if (cause instanceof IOException && "Connection reset by peer".equals(cause.getMessage())) {
+                    //it's client closed connection already when we tried to send a message back - ignore
+                } else {
+                    fail();
+                }
+            }
+        };
+
+        ChannelInboundHandler trackingHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                if (cause instanceof ReadTimeoutException) {
+                    timeoutGotLatch.countDown();
+                } else {
+                    unexpectedExceptionInClientHandler.set(true);
+                }
+            }
+        };
+
+        Channel server = newServer(pongAfterTwoMessagesHandler);
+        Channel client = newClient(server.localAddress(), readAfterWriteTimeoutHandler, trackingHandler);
+        try {
+            client.writeAndFlush(newOneMessage());
+            Thread.sleep(TIMEOUT_MS / 10);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)